packages/grid/src/styles/vaadin-grid-base-styles.js

```diff
@@ -189,145 +159,146 @@ export const gridStyles = css`
     position: relative;
     align-items: center;
     white-space: nowrap;
-    border-inline-start: var(--_column-border-width) solid var(--_border-color);
   }
 
-  [part~='first-column-cell'] {
-    border-inline-start: 0;
-  }
+  /* Row and cell borders */
 
-  [part~='first-header-row-cell'],
-  [part~='first-footer-row-cell'],
-  [part~='first-row-cell'] {
-    margin-top: 0;
-    border-top: 0;
+  [part~='last-header-row']::before,
+  [part~='first-footer-row']::before {
+    position: absolute;
+    inset-inline: 0;
+    border-block: var(--_row-border-width) var(--_border-color);
+    transform: translateX(var(--_grid-horizontal-scroll-position));
   }
 
-  table[has-header] [part~='first-row-cell'] {
-    border-top: var(--_row-border-width) solid var(--_border-color);
+  :host([overflow~='top']) [part~='last-header-row']::before {
+    content: '';
+    inset-block-end: calc(var(--_row-border-width) * -1);
+    border-block-end-style: solid;
   }
 
-  [part~='last-column-cell'] {
-    --_end: 1;
+  :host([overflow~='bottom']) [part~='first-footer-row']::before,
+  :host(:not([overflow~='top'])) #scroller:not([empty-state]) [part~='first-footer-row']::before {
+    content: '';
+    inset-block-start: calc(var(--_row-border-width) * -1);
+    border-block-start-style: solid;
   }
 
-  [part~='last-column-cell']:is([part~='header-cell'], [part~='footer-cell']) {
-    --_end-opaque: 1;
+  [part~='body-row'] {
+    scroll-margin-block-end: var(--_row-border-width);
   }
 
-  [part~='last-row-cell']:where(:not([part~='details-opened-row-cell'])),
-  [part~='last-footer-row-cell'] {
-    border-bottom: 0;
-    --_bottom: 1;
+  /* stylelint-disable-next-line no-duplicate-selectors */
+  [part~='cell'] {
+    border-block: var(--_row-border-width) var(--_border-color);
+    border-inline: var(--_column-border-width) var(--_border-color);
   }
 
-  [part~='last-frozen-cell'] {
-    --_end: 1;
+  [part~='header-cell']:not([part~='first-header-row-cell']),
+  [part~='footer-cell']:not([part~='first-footer-row-cell']),
+  [part~='body-cell']:not([part~='first-row-cell']),
+  #table[has-header] [part~='first-row-cell'] {
+    border-block-start-style: solid;
   }
 
-  [part~='last-frozen-cell'] + [part~='cell'] {
-    border-inline-start-color: transparent;
+  [part~='details-opened-row-cell'],
+  #table[has-footer] [part~='last-row-cell'],
+  #table[has-footer] [part~='last-row'] [part~='details-cell'],
+  :host(:not([overflow~='bottom']):not([overflow~='top'])) [part~='last-row-cell'],
+  :host(:not([overflow~='bottom']):not([overflow~='top'])) [part~='last-row'] [part~='details-cell'] {
+    border-block-end-style: solid;
   }
 
-  [part~='first-frozen-to-end-cell'] {
-    border-inline-start: 0;
-    --_start: 1;
+  [part~='header-cell']:not([part~='first-column-cell']),
+  [part~='footer-cell']:not([part~='first-column-cell']),
+  [part~='body-cell']:not([part~='first-column-cell']) {
```

Comment:
Can we simplify this?

```suggestion
  :is(th, td):not([part~='first-column-cell']) {
```

Reply:
Thanks, I'm actually considering trying to improve selectors in a follow-up PR to reduce the scope here. Would that work?

Reply:
Yeah, that’s fine.